internal/internal_workflow.go

```diff
@@ -972,20 +974,53 @@ func getStackTrace(coroutineName, status string, stackDepth int) string {
 
 func getStackTraceRaw(top string, omitTop, omitBottom int) string {
 	stack := stackBuf[:runtime.Stack(stackBuf[:], false)]
-	rawStack := strings.TrimRightFunc(string(stack), unicode.IsSpace)
+	outStack := filterStackTrace(string(stack), omitTop, omitBottom)
+	return strings.Join([]string{top, outStack}, "\n")
+}
+
+func filterStackTrace(stack string, omitTop, omitBottom int) string {
+	stack = strings.TrimRightFunc(stack, unicode.IsSpace)
 	if disableCleanStackTraces {
-		return rawStack
+		return stack
 	}
-	lines := strings.Split(rawStack, "\n")
+
+	lines := strings.Split(stack, "\n")
 	omitEnd := len(lines) - omitBottom
 	// If the start is after the end, the depth was invalid originally so return
 	// the entire raw stack
 	if omitTop > omitEnd {
-		return rawStack
+		return stack
+	}
+	return strings.Join(lines[omitTop:omitEnd], "\n")
+}
+
+func getCoroStackTrace(crt *coroutineState, status string, stackDepth int) string {
+	// Can't dump goroutines selectively :(
+	// Instead, we identify a coroutine's stack trace by the *coroutineState pointer value
+	// in its function arguments. To avoid false positives, we also match on the fixed
+	// member function name.
+	stacks := stackBuf[:runtime.Stack(stackBuf[:], true)]
+	coros := bytes.Split(stacks, []byte("\n\n"))
+
+	var coroStack []byte
+	needle := []byte(fmt.Sprintf("/internal.(*coroutineState).run(%p,", crt))
+	for _, c := range coros {
+		if bytes.Contains(c, needle) {
+			coroStack = c
+			break
+		}
+	}
+	if coroStack == nil {
```

Comment:
nit: would prefer these two situations be errors returned and handled by the caller

Reply:
I will change that. How would you like me to handle that error inside the deadlock detector? Format the error into the stack trace (similar to how the the return value looks now)?

Ignoring it seems like a bad idea, and I can't log it since `*coroutineState` doesn't have access to the environment's logger.

Reply:
>How would you like me to handle that error inside the deadlock detector? Format the error into the stack trace (similar to how the the return value looks now)?

Yes, formatting it into the stack trace is fine IMO.